src/services/expenses.ts

```diff
@@ -0,0 +1,113 @@
+'use strict'
+
+import { ExpensesStorage } from '../storages'
+import { ServiceRegistry } from './serviceRegistry'
+import { isValidDateString } from '../shared/validation'
+import { ensure } from '../utils'
+import type {
+  RawExpensesFilter,
+  ExpensesFilter,
+  ExpensesInput
+} from '../types'
+
+export default class ExpensesService {
+  private constructor (
+    private storage: ExpensesStorage,
+    private serviceRegistry: ServiceRegistry
+  ) {
+    serviceRegistry.register('ExpensesService', this)
+  }
+
+  public static create (
+    storage: ExpensesStorage,
+    serviceRegistry: ServiceRegistry
+  ) {
+    return new ExpensesService(storage, serviceRegistry)
+  }
+
+  public async getAll (filter: RawExpensesFilter) {
+    this.validateExpenseFilter(filter)
+
+    return this.storage.findMany({
+      select: {
+        id: true,
+        userId: true,
+        spentAt: true,
+        title: true,
+        amount: true,
+        category: true,
+        note: true
+      },
+      where: {
+        userId: { equals: filter.userId },
+        spentAt: { gte: filter.from, lte: filter.to },
+        category: { in: filter.categories }
+      }
+    })
```

Comment:
Looks like storage exposes too many low-level details.

Reply:
Got it

I guess we need some type for a filter of each entity or even better, a general filter parser to more low-level things (or both)